File: sdk/typescript/tests/table-data.test.ts
import { TableData, Column, ColumnType } from '../src/table-data';

describe('TableData', () => {
  describe('constructor', () => {
    it('should create empty table with no arguments', () => {
      const table = new TableData();
      expect(table.rowCount()).toBe(0);
      expect(table.columnCount()).toBe(0);
      expect(table.isEmpty()).toBe(true);
    });

    it('should create table with columns and rows', () => {
      const columns: Column[] = [
        { name: 'id', type: ColumnType.INTEGER, nullable: false },
        { name: 'name', type: ColumnType.STRING, nullable: true },
      ];
      const rows = [
        [1, 'Alice'],
        [2, 'Bob'],
      ];

      const table = new TableData(columns, rows);
      expect(table.rowCount()).toBe(2);
      expect(table.columnCount()).toBe(2);
      expect(table.isEmpty()).toBe(false);
    });
  });

  describe('addRow', () => {
    it('should add a single row', () => {
      const table = new TableData();
      table.addRow([1, 'test']);
      expect(table.rowCount()).toBe(1);
    });

    it('should add multiple rows individually', () => {
      const table = new TableData();
      table.addRow([1, 'first']);
      table.addRow([2, 'second']);
      expect(table.rowCount()).toBe(2);
    });
  });

  describe('addRows', () => {
    it('should add multiple rows at once', () => {
      const table = new TableData();
      table.addRows([
        [1, 'first'],
        [2, 'second'],
        [3, 'third'],
      ]);
      expect(table.rowCount()).toBe(3);
    });
  });

  describe('getRow', () => {
    it('should return row at index', () => {
      const table = new TableData();
      table.addRow([1, 'test']);
      const row = table.getRow(0);
      expect(row).toEqual([1, 'test']);
    });

    it('should return undefined for invalid index', () => {
      const table = new TableData();
      expect(table.getRow(0)).toBeUndefined();
      expect(table.getRow(-1)).toBeUndefined();
    });
  });

  describe('getCell', () => {
    it('should return cell value', () => {
      const table = new TableData();
      table.addRow([1, 'test']);
      expect(table.getCell(0, 0)).toBe(1);
      expect(table.getCell(0, 1)).toBe('test');
    });

    it('should return undefined for invalid indices', () => {
      const table = new TableData();
      table.addRow([1, 'test']);
      expect(table.getCell(1, 0)).toBeUndefined();
      expect(table.getCell(0, 2)).toBeUndefined();
    });
  });

  describe('clear', () => {
    it('should remove all rows', () => {
      const table = new TableData();
      table.addRows([
        [1, 'first'],
        [2, 'second'],
      ]);
      expect(table.rowCount()).toBe(2);
      
      table.clear();
      expect(table.rowCount()).toBe(0);
      expect(table.isEmpty()).toBe(true);
    });
  });

  describe('columns', () => {
    it('should get and set columns', () => {
      const table = new TableData();
      const columns: Column[] = [
        { name: 'id', type: ColumnType.INTEGER },
        { name: 'value', type: ColumnType.FLOAT },
      ];

      table.setColumns(columns);
      expect(table.getColumns()).toEqual(columns);
      expect(table.columnCount()).toBe(2);
    });
  });
});

File: sdk/go/table_data_test.go
package compliance

import (
	"testing"
)

func TestTableDataCreation(t *testing.T) {
	columns := []ColumnMetadata{
		{Name: "id", Type: "INTEGER", Nullable: false},
		{Name: "name", Type: "STRING", Nullable: true},
	}
	table := NewTableData(columns)

	if table.ColumnCount() != 2 {
		t.Errorf("Expected 2 columns, got %d", table.ColumnCount())
	}
	if !table.IsEmpty() {
		t.Error("Expected table to be empty")
	}
}

func TestTableDataAddRow(t *testing.T) {
	table := NewTableData([]ColumnMetadata{
		{Name: "id", Type: "INTEGER"},
	})

	table.AddRow(Row{1})
	table.AddRow(Row{2})

	if table.RowCount() != 2 {
		t.Errorf("Expected 2 rows, got %d", table.RowCount())
	}
	if table.IsEmpty() {
		t.Error("Expected table to not be empty")
	}
}

func TestTableDataGetCell(t *testing.T) {
	table := NewTableData([]ColumnMetadata{
		{Name: "value", Type: "INTEGER"},
	})
	table.AddRow(Row{42})

	cell, err := table.GetCell(0, 0)
	if err != nil {
		t.Fatalf("GetCell failed: %v", err)
	}

	if cell != 42 {
		t.Errorf("Expected cell value 42, got %v", cell)
	}
}

func TestTableDataGetCellOutOfBounds(t *testing.T) {
	table := NewTableData([]ColumnMetadata{
		{Name: "value", Type: "INTEGER"},
	})
	table.AddRow(Row{42})

	_, err := table.GetCell(1, 0)
	if err == nil {
		t.Error("Expected error for out of bounds row index")
	}

	_, err = table.GetCell(0, 1)
	if err == nil {
		t.Error("Expected error for out of bounds column index")
	}
}

func TestTableDataClear(t *testing.T) {
	table := NewTableData([]ColumnMetadata{
		{Name: "id", Type: "INTEGER"},
	})
	table.AddRow(Row{1})
	table.AddRow(Row{2})

	table.Clear()

	if !table.IsEmpty() {
		t.Error("Expected table to be empty after clear")
	}
	if table.RowCount() != 0 {
		t.Errorf("Expected 0 rows after clear, got %d", table.RowCount())
	}
}

func TestIsNull(t *testing.T) {
	if !IsNull(nil) {
		t.Error("Expected nil to be null")
	}

	if IsNull(42) {
		t.Error("Expected 42 to not be null")
	}

	if IsNull("test") {
		t.Error("Expected 'test' to not be null")
	}
}

func TestCellValueToString(t *testing.T) {
	tests := []struct {
		value    CellValue
		expected string
	}{
		{nil, "NULL"},
		{42, "42"},
		{"hello", "hello"},
		{3.14, "3.14"},
		{true, "true"},
	}

	for _, tt := range tests {
		result := CellValueToString(tt.value)
		if result != tt.expected {
			t.Errorf("Expected '%s', got '%s'", tt.expected, result)
		}
	}
}

